### ISSUE TITLE
How to generate Firestore security rules?

### ISSUE THREAD

**Issue by jmagaram** [DZakh/rescript-struct#34]
I'm using this library. Very cool! After I define a schema, like a string that is a minimum of 4 characters and a max of 10 characters, matching a regex pattern, I want to generate a Firebase Firestore security rules file that ensures data in the db matches this definition. The rules file has text in it like...
```
data.size() >= 4 && data.size()<=10 && data.matches("[a-z]+")
```
How would I go about converting a struct here to this kind of text?

**Comment by DZakh** [comment]
Am I right that you want to take a string struct that uses built-in refinements like this:
```
let struct = S.string()->S.String.min(4)->S.String.max(10)->S.String.pattern(%re("[a-z]+"))
```
And then use it to generate a string with Firestore security rules?
```
Console.log(Firestore.SecurityRules.make(~struct))
// Output: "data.size() >= 4 && data.size()<=10 && data.matches(\"[a-z]+\")"
```

**Comment by jmagaram** [comment]
Yes but not just for strings. If I'm using the built in .min(3), .max(9), regex standard patterns maybe I can do most of it automatically as long as I can **see inside the struct**. FireStore lets you do validation on "objects" so I'll need to put together conditions on each field, like field1requirements && field2requirements etc. But if there is custom code in some field I'd like to know that and will probably have to bail/fail or do something custom.

**Comment by DZakh** [comment]
Can you show an example of some `FireStore` validation object?

**Comment by jmagaram** [comment]
Can't do it now but here is a tool  that does it from zod. 

https://github.com/yarnaimo/fireschema

The security rules are in a text file with a bunch of stuff unrelated to ReScriptStruct. But there are places in the file where you'll write "allow read, write : if (conditions)" where the condition is based on the rules for what data is allowed. 

https://firebase.google.com/docs/firestore/security/rules-structure

**Comment by jmagaram** [comment]
Here is a tool where you start from a schema similar to a TypeScript and it generates real TypeScript types AND the security rule file so you can keep the two in sync. 

https://github.com/bijoutrouvaille/fireward

**Comment by jmagaram** [comment]
Some of the rule conditions are based on authentication. Like this (not syntactically correct but you'll get the idea...

allow read : if not(user is anonymous) && user.role=="admin" && (insert the data rules here)

**Comment by DZakh** [comment]
Oh, now it's super clear to me. Yes, it's possible to do with ReScript struct design. But I've scoped out the feature of getting the information about refinements usage, so for the version `3.1.0` some security rules are not possible:

<img width="860" alt="image" src="https://user-images.githubusercontent.com/49292466/222775646-43559cdf-92be-4491-84ba-0922628a1615.png">

It's easy to add. I'm preparing the next `rescript-struct` release in one-two week. I'll also add the information about refinements, so it'll be possible to output all security rules from the screenshot.

As for the implementation of the converter of rescript-structs to security rules, you can take a look at https://github.com/DZakh/rescript-json-schema/blob/9005594ee3ddd10af70d5543513324637b4ae978/src/JSONSchema.res#L156, it's very similar to what you need.

**Comment by DZakh** [comment]
Here's the simplified version of possible code:
```

let makeSecurityRule = (~localtion=[], struct) => {
  switch struct->S.classify {
  | S.Unknown => "true"
  | S.String => `data${localtion->Location.toSecurityRulePath} is string`
  | S.Bool => `data${localtion->Location.toSecurityRulePath} is bool`
  | S.Float => `data${localtion->Location.toSecurityRulePath} is number`
  | S.Int => `data${localtion->Location.toSecurityRulePath} is int`
  | S.Tuple(structs) => `(data${localtion->Location.toSecurityRulePath} is list && ${structs
      ->Array.mapWithIndex((itemStruct, index) =>
        makeSecurityRule(~location=location->Array.concat([index], itemStruct))
      )
      ->Array.joinWith(" && ")})`
  }
}
```

**Comment by jmagaram** [comment]
Awesome! I think it is possible to write custom JavaScript logic for validation in ReScriptStruct. Is that correct? Will I be able to know if such a custom rule was written? Because those rules can never be automatically translated to Firestore and I'd want to throw and exception or something.

**Comment by jmagaram** [comment]
Actually I assume the custom code refinement might fall into the Unknown category.

**Comment by DZakh** [comment]
> https://github.com/DZakh/rescript-struct/issues/34#issuecomment-1453809361

I've created the PR to solve the problem https://github.com/DZakh/rescript-struct/pull/39

>  I think it is possible to write custom JavaScript logic for validation in ReScriptStruct. Is that correct?

Yes, it is.

> Actually, I assume the custom code refinement might fall into the Unknown category.

Yes, but if somebody uses `S.transform`, or `S.refine` the struct tag won't be changed to Unknown

> Will I be able to know if such a custom rule was written?

It's not possible right now. As I know, it's also not possible with Zod. I assume they just ignore structs with custom validation logic. But if not, I can add an ability to check whether there's custom validation logic applied.

**Comment by jmagaram** [comment]
This looks really good. I can get the refinement kind for each type like int float array and string. So I can see that a string has a minimum length and maximum length for example.

Can I get the larger structure? Like can I know that a struct refers to an "object" with 3 fields - "firstName" string "lastName" string and "age" int?

And can I know if a field is optional?

It may be ok without it. The place where I define the struct I can define the security rules and pull out the metadata without having to "repeat myself". So if I know a field is a primitive I'll use the metadata to get the details. But for the larger picture like what fields are in the struct maybe I'll have to do that manually. This would only work for a struct I built myself where I know what it looks like. 

I haven't tried generating security rules yet. Probably a few weeks away.

**Comment by jmagaram** [comment]
Maybe the Tuple is the "object" so I scan it recursively? Optional fields are important. Can I know about that?

**Comment by DZakh** [comment]
> Can I get the larger structure? Like can I know that a struct refers to an "object" with 3 fields - "firstName" string "lastName" string and "age" int?

Yes, it's already possible in the current version https://github.com/DZakh/rescript-json-schema/blob/9005594ee3ddd10af70d5543513324637b4ae978/src/JSONSchema.res#L205

> And can I know if a field is optional?

Yes, it's already possible in the current version https://github.com/DZakh/rescript-json-schema/blob/9005594ee3ddd10af70d5543513324637b4ae978/src/JSONSchema.res#L150

**Comment by jmagaram** [comment]
Great. One last thing. Can I determine if something is a choice discriminated union? I use those everywhere. It's an A or B or C.

**Comment by DZakh** [comment]
Yep, https://github.com/DZakh/rescript-json-schema/blob/9005594ee3ddd10af70d5543513324637b4ae978/src/JSONSchema.res#L187

**Comment by DZakh** [comment]
I've released rescript-struct@4.0.0 with the ability to get information about refinements. You can find a usage example here https://github.com/DZakh/rescript-struct/blob/57b6aa932207c2b27070b37df6d90503afbcfe9f/src/S.res#L3628-L3660

Feel free to reopen/open a new issue if you have any questions or feature requests.
